### ISSUE TITLE
AwsCustomResource: DescribeVpcEndpointServices API not returning information

### ISSUE THREAD

**Issue by seanwynwins** [aws/aws-cdk#23601]
### Describe the bug

When I deploy my stack with an AWS Custom Resource for the DescribeVpcEndpointServices API with the parameter "ServiceNames", I receive the following:

Stack Deployments Failed: Error: The stack named <> failed to deploy: UPDATE_ROLLBACK_COMPLETE: CustomResource attribute error: Vendor response doesn't contain ServiceDetails key

### Expected Behavior

I should have been able to get the service details of the vpc endpoint from this AWS custom resource. I was able to have another EC2 API (describeNetworkInterfaces) work with the same format, just replaced the action with describeNetworkInterfaces and the parameter name. The VPC endpoint I used is definitely valid, I was able to call the API through the AWS CLI and a manual lambda I made.

### Current Behavior

6:17:25 PM | CREATE_FAILED        | AWS::EC2::Subnet                          | ...Subnet08F0B33B
CustomResource attribute error: Vendor response doesn't contain ServiceDetails key in object arn:aws:cloudformation:us-east-1:734789510895:stack/beta/05874f40-8de
e-11ed-b85a-12795384352b|DescribeVpcEndpointServicesA596F9B2|82de965e-bf1b-4217-95bd-58f45b6f6206 in S3 bucket cloudformation-custom-resource-storage-useast1

### Reproduction Steps

```
  const describeVpcEndpointServices = new custom_resources.AwsCustomResource(
      this,
      "DescribeVpcEndpointServices",
      {
        onCreate: {
          service: "EC2",
          action: "describeVpcEndpointServices",
          parameters: {
            ServiceNames: [com.amazonaws.us-east-1.voiceid],
          },
          physicalResourceId: custom_resources.PhysicalResourceId.of(Date.now().toString()),
        },
        onUpdate: {
          service: "EC2",
          action: "describeVpcEndpointServices",
          parameters: {
            ServiceNames: [com.amazonaws.us-east-1.voiceid],
          },
          physicalResourceId: custom_resources.PhysicalResourceId.of(Date.now().toString()),
        },
        policy: {
          statements: [
            new iam.PolicyStatement({
              actions: ["*"],
              resources: ["*"],
            }),
          ],
        },
        functionName: "DescribeVpcEndpointServices"
      }
    );

    const responseString = describeVpcEndpointServices.getResponseField("ServiceDetails");
```

The last line returns an error.

### CDK CLI Version

no cdk version found, output that it was using latest version

### Node.js Version

v14.21.2

### OS

Linux

### Language

Typescript

**Comment by peterwoodworth** [comment]
Hey @seanwynwins,

This is occurring because `ServiceDetails` isn't a concrete single value, but instead an array. CloudFormation doesn't allow you to refer to values returned by custom resources which aren't a primitive value type.

I'm not sure exactly what you're trying to do, but you may need to come up with another way of doing what you're trying to do. Let me know the specifics and I can try to help 🙂

**Comment by seanwynwins** [comment]
Thanks, I had gotten help from someone who looked into the source code and saw that only primitive values can be returned. I was actually trying to get `ServiceDetails.0.AvailabilityZones` which wasn't working. Could this be made more clear in the [documentation](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources.AwsCustomResource.html), as myself and multiple people were unable to debug this for a while haha

**Comment by seanwynwins** [comment]
And also, I need to get a list of availability zones of a vpc endpoint service, which I'm unable to do since I can't iterate over the array. Are there any implementations that could help with this?

**Comment by peterwoodworth** [comment]
> And also, I need to get a list of availability zones of a vpc endpoint service, which I'm unable to do since I can't iterate over the array. Are there any implementations that could help with this?

This is tricky, and there's not a super great way to do this. The simplest way would be something like this:

First, write a shell script that retrieves the list and writes it to a JSON file. Then read that JSON file in your cdk app and do the loop there

The problems with this solution are that the list cannot be deployed as part of your app - it has to be deployed before you synth. Additionally, if the list ever changes, you need to manually invalidate it. And if you need to do this in multiple accounts, you need to do have credentials to all of them during synth

**Comment by seanwynwins** [comment]
Would making a custom resource for this work? I've tried doing so but am running into some issues. I made a custom lambda to return the array of availability zones and use them as an argument in creating a VPC but I'm unable to get it to work. 

Using `customResource.getAtt('AvailabilityZones').toStringList()`

doesn't seem to work

**Comment by peterwoodworth** [comment]
I don't think that would work - The reason you cannot iterate is because the way you retrieve the actual data is by referencing it in the template. By the time the custom resource starts running you've already created and deployed your template, which means you need to have already known the size of the array before the custom resource could return that information.

**Comment by seanwynwins** [comment]
i don't need to iterate. i just need to plug in the returned availability zones array into an argument for creating a vpc. I thought custom resource allowed you to return lists? am I misunderstanding the "toStringList" method?

**Comment by seanwynwins** [comment]
if the "toStringList" method is unable to return a list that can actually be plugged into something, this also has to be made clear...

**Comment by seanwynwins** [comment]
```
vpcEndpointAvailabilityZones = customResource.getAtt('AvailabilityZones').toStringList()
    const vpc = new ec2.Vpc(this, 'vpc', {
      cidr: "172.31.0.0/16",
      availabilityZones: vpcEndpointAvailabilityZones,
    })
```

this returns an error of "Resolution error: Found an encoded list token string in a scalar string context."

**Comment by peterwoodworth** [comment]
This won't be an option - at the end of the day when referencing a return value from a custom resource, you'll need to have specified a spot in the array within the template. `toStringList()` works with tokens that can resolve as arrays, and this token cannot resolve as an array due to the specific kind of reference it is.

**Comment by CodeSammich** [comment]
For future readers, OP and I are working on the same project. We had to ultimately hard-code the number of AZs available and do a loop like this where `numVpcEndpointServiceAzs` is a variable we hard-coded to 3 for our use case in the region.

```
    const vpcEndpointServiceAZs :string[] = [];
    for (let i = 0; i < numVpcEndpointServiceAzs; i++) {
      vpcEndpointServiceAZs.push(vpcEndpointAvailabilityZonesApi.getResponseField(`ServiceDetails.0.AvailabilityZones.${i}`))    
    }
    const myVPC = new ec2.Vpc(this, 'myVPC', {
      cidr: "172.31.0.0/16",
      availabilityZones: vpcEndpointServiceAZs,
      vpcName: "myVPC"
    })
```
